File: examples/pubstress.c
/*
 * Copyright (c) 2004-2010 The Trustees of Indiana University and Indiana
 *                         University Research and Technology
 *                         Corporation.  All rights reserved.
 * Copyright (c) 2004-2011 The University of Tennessee and The University
 *                         of Tennessee Research Foundation.  All rights
 *                         reserved.
 * Copyright (c) 2004-2005 High Performance Computing Center Stuttgart,
 *                         University of Stuttgart.  All rights reserved.
 * Copyright (c) 2004-2005 The Regents of the University of California.
 *                         All rights reserved.
 * Copyright (c) 2006-2013 Los Alamos National Security, LLC.
 *                         All rights reserved.
 * Copyright (c) 2009-2012 Cisco Systems, Inc.  All rights reserved.
 * Copyright (c) 2011      Oak Ridge National Labs.  All rights reserved.
 * Copyright (c) 2013-2020 Intel, Inc.  All rights reserved.
 * Copyright (c) 2015      Mellanox Technologies, Inc.  All rights reserved.
 * Copyright (c) 2019      IBM Corporation.  All rights reserved.
 * Copyright (c) 2021-2025 Nanook Consulting  All rights reserved.
 * $COPYRIGHT$
 *
 * Additional copyrights may follow
 *
 * $HEADER$
 *
 */

#define _GNU_SOURCE
#include <stdbool.h>
#include <stdio.h>
#include <string.h>
#include <stdlib.h>
#include <time.h>
#include <unistd.h>
#include <libgen.h>

#include <pmix.h>

#include "examples.h"

#define NITER 10

int main(int argc, char **argv)
{
    pmix_proc_t myproc;
    int rc, n;
    pmix_value_t *val = NULL;
    pmix_proc_t proc;
    uint32_t nprocs;
    pmix_info_t info[2], lkinfo[2];
    pmix_pdata_t pdata;
    pmix_persistence_t firstread = PMIX_PERSIST_FIRST_READ;
    int timeout=10;
    char *tmp;
    int iters;
    size_t ninfo, bcount=0;
    char *keys[3];

    if (1 < argc) {
        iters = strtol(argv[1], NULL, 10);
    } else {
        iters = NITER;
    }

    /* init us */
    if (PMIX_SUCCESS != (rc = PMIx_Init(&myproc, NULL, 0))) {
        fprintf(stderr, "Client ns %s rank %d: PMIx_Init failed: %d\n", myproc.nspace, myproc.rank,
                rc);
        exit(0);
    }

    /* get our job size */
    PMIX_PROC_CONSTRUCT(&proc);
    PMIX_LOAD_PROCID(&proc, myproc.nspace, PMIX_RANK_WILDCARD);
    if (PMIX_SUCCESS != (rc = PMIx_Get(&proc, PMIX_JOB_SIZE, NULL, 0, &val))) {
        fprintf(stderr, "Client ns %s rank %d: PMIx_Get job size failed: %d\n", myproc.nspace,
                myproc.rank, rc);
        goto done;
    }
    nprocs = val->data.uint32;
    PMIX_VALUE_RELEASE(val);
    // require at least 4 procs
    if (nprocs < 4) {
        if (0 == myproc.rank) {
            fprintf(stderr, "%s requires at least 4 processes\n", basename(argv[0]));
        }
        exit(1);
    }

    PMIX_INFO_LOAD(&info[1], PMIX_PERSISTENCE, &firstread, PMIX_PERSIST);
    PMIX_INFO_LOAD(&lkinfo[0], PMIX_WAIT, NULL, PMIX_BOOL);
    PMIX_INFO_LOAD(&lkinfo[1], PMIX_TIMEOUT, &timeout, PMIX_INT);

    if (myproc.rank < 2) {
        ninfo = 2;
        for (n=0; n < iters; n++) {
            if (0 == myproc.rank) {
                if (n % 2) { // delay us a bit so other ranks get ahead
                    sleep(1);
                }
                /* publish something */
                if (0 > asprintf(&tmp, "FOOBAR:%s.%u:%d", myproc.nspace, myproc.rank, n)) {
                    goto done;
                }
                PMIX_INFO_LOAD(&info[0], tmp, &n, PMIX_INT);
                free(tmp);
                if (PMIX_SUCCESS != (rc = PMIx_Publish(info, ninfo))) {
                    fprintf(stderr, "Client ns %s rank %d: PMIx_Publish failed: %d\n", myproc.nspace,
                            myproc.rank, rc);
                    goto done;
                }
                PMIX_INFO_DESTRUCT(&info[0]);

                /* lookup other rank's value */
                PMIX_PDATA_CONSTRUCT(&pdata);
                if (0 > asprintf(&tmp, "BAZ:%s.%u:%d", myproc.nspace, 1, n)) {
                    goto done;
                }
                PMIX_LOAD_KEY(pdata.key, tmp);
                free(tmp);
                // lookup value
                if (PMIX_SUCCESS != (rc = PMIx_Lookup(&pdata, 1, lkinfo, 2))) {
                    fprintf(stderr, "Client ns %s rank %d: PMIx_Lookup failed: %d\n", myproc.nspace,
                            myproc.rank, rc);
                    goto done;
                }
                /* check the return for value and source */
                if (0 != strncmp(myproc.nspace, pdata.proc.nspace, PMIX_MAX_NSLEN)) {
                    fprintf(stderr, "Client ns %s rank %d: PMIx_Lookup returned wrong nspace: %s\n",
                            myproc.nspace, myproc.rank, pdata.proc.nspace);
                    goto done;
                }
                if (1 != pdata.proc.rank) {
                    fprintf(stderr, "Client ns %s rank %d: PMIx_Lookup returned wrong rank: %d\n",
                            myproc.nspace, myproc.rank, pdata.proc.rank);
                    goto done;
                }
                if (PMIX_INT != pdata.value.type) {
                    fprintf(stderr, "Client ns %s rank %d: PMIx_Lookup returned wrong type: %d\n",
                            myproc.nspace, myproc.rank, pdata.value.type);
                    goto done;
                }
                if (n != pdata.value.data.integer) {
                    fprintf(stderr, "Client ns %s rank %d: PMIx_Lookup returned wrong value: %d\n",
                            myproc.nspace, myproc.rank, (int) pdata.value.data.uint8);
                    goto done;
                }
                PMIX_PDATA_DESTRUCT(&pdata);
                fprintf(stderr, "PUBLISH-LOOKUP SUCCEEDED: %d\n", n);
                ninfo = 2;
            } else if (1 == myproc.rank) {
                if (0 == n) {
                    sleep(1);
                }
                /* lookup other rank's value */
                PMIX_PDATA_CONSTRUCT(&pdata);
                if (0 > asprintf(&tmp, "FOOBAR:%s.%u:%d", myproc.nspace, 0, n)) {
                    goto done;
                }
                PMIX_LOAD_KEY(pdata.key, tmp);
                free(tmp);
                // check value
                if (PMIX_SUCCESS != (rc = PMIx_Lookup(&pdata, 1, lkinfo, 2))) {
                    fprintf(stderr, "Client ns %s rank %d: PMIx_Lookup failed: %d\n", myproc.nspace,
                            myproc.rank, rc);
                    goto done;
                }
                /* check the return for value and source */
                if (0 != strncmp(myproc.nspace, pdata.proc.nspace, PMIX_MAX_NSLEN)) {
                    fprintf(stderr, "Client ns %s rank %d: PMIx_Lookup returned wrong nspace: %s\n",
                            myproc.nspace, myproc.rank, pdata.proc.nspace);
                    goto done;
                }
                if (0 != pdata.proc.rank) {
                    fprintf(stderr, "Client ns %s rank %d: PMIx_Lookup returned wrong rank: %d\n",
                            myproc.nspace, myproc.rank, pdata.proc.rank);
                    goto done;
                }
                if (PMIX_INT != pdata.value.type) {
                    fprintf(stderr, "Client ns %s rank %d: PMIx_Lookup returned wrong type: %d\n",
                            myproc.nspace, myproc.rank, pdata.value.type);
                    goto done;
                }
                if (n != pdata.value.data.integer) {
                    fprintf(stderr, "Client ns %s rank %d: PMIx_Lookup returned wrong value: %d\n",
                            myproc.nspace, myproc.rank, (int) pdata.value.data.uint8);
                    goto done;
                }
                PMIX_PDATA_DESTRUCT(&pdata);
                /* publish something */
                if (0 > asprintf(&tmp, "BAZ:%s.%u:%d", myproc.nspace, myproc.rank, n)) {
                    goto done;
                }
                PMIX_INFO_LOAD(&info[0], tmp, &n, PMIX_INT);
                free(tmp);
                if (PMIX_SUCCESS != (rc = PMIx_Publish(info, 2))) {
                    fprintf(stderr, "Client ns %s rank %d: PMIx_Publish failed: %d\n", myproc.nspace,
                            myproc.rank, rc);
                    goto done;
                }
                PMIX_INFO_DESTRUCT(&info[0]);
            }
        }
        goto done;
    }

    // we cannot participate in the "firstread" cases as only
    // one other proc can read the value
    /* lookup other rank's value */
    for (n=0; n < iters; n++) {
        if (2 == myproc.rank) {
            if (n % 2) { // delay us a bit so other ranks get ahead
                sleep(1);
            }
            // publish something
            if (0 > asprintf(&tmp, "BIGTEST:%s.%u:%d", myproc.nspace, myproc.rank, n)) {
                goto done;
            }
            bcount = 1234 + n;
            PMIX_INFO_LOAD(&info[0], tmp, &bcount, PMIX_SIZE);
            free(tmp);
                fprintf(stderr, "Client ns %s rank %d: Published: %lu\n", myproc.nspace,
                        myproc.rank, (unsigned long)bcount);
            if (PMIX_SUCCESS != (rc = PMIx_Publish(info, 1))) {
                fprintf(stderr, "Client ns %s rank %d: PMIx_Publish failed: %d\n", myproc.nspace,
                        myproc.rank, rc);
                goto done;
            }
            PMIX_INFO_DESTRUCT(&info[0]);
        } else {
            // everyone else retrieves it
            PMIX_PDATA_CONSTRUCT(&pdata);
            if (0 > asprintf(&tmp, "BIGTEST:%s.%u:%d", myproc.nspace, 2, n)) {
                goto done;
            }
            PMIX_LOAD_KEY(pdata.key, tmp);
            free(tmp);
            // check value
            if (PMIX_SUCCESS != (rc = PMIx_Lookup(&pdata, 1, lkinfo, 2))) {
                fprintf(stderr, "Client ns %s rank %d: PMIx_Lookup failed: %d\n", myproc.nspace,
                        myproc.rank, rc);
                goto done;
            }
            /* check the return for value and source */
            if (!PMIX_CHECK_NSPACE(myproc.nspace, pdata.proc.nspace)) {
                fprintf(stderr, "Client ns %s rank %d: PMIx_Lookup returned wrong nspace: %s\n",
                        myproc.nspace, myproc.rank, pdata.proc.nspace);
                goto done;
            }
            if (2 != pdata.proc.rank) {
                fprintf(stderr, "Client ns %s rank %d: PMIx_Lookup returned wrong rank: %d\n",
                        myproc.nspace, myproc.rank, pdata.proc.rank);
                goto done;
            }
            if (PMIX_SIZE != pdata.value.type) {
                fprintf(stderr, "Client ns %s rank %d: PMIx_Lookup returned wrong type: %s\n",
                        myproc.nspace, myproc.rank, PMIx_Data_type_string(pdata.value.type));
                goto done;
            }
            bcount = pdata.value.data.size;
            if (bcount != (size_t)(1234 + n)) {
                fprintf(stderr, "Client ns %s rank %d: PMIx_Lookup returned wrong value: %lu\n",
                        myproc.nspace, myproc.rank, (unsigned long)bcount);
                goto done;
            }
            PMIX_PDATA_DESTRUCT(&pdata);
            fprintf(stderr, "Client ns %s rank %d: PUBLISH-LOOKUP BIGTEST SUCCEEDED: %d\n",
                    myproc.nspace, myproc.rank, n);
        }
    }

    PMIX_LOAD_PROCID(&proc, myproc.nspace, PMIX_RANK_WILDCARD);
    rc = PMIx_Fence(&proc, 1, NULL, 0);
    if (2 == myproc.rank) {
        if (0 > asprintf(&keys[0], "BIGTEST:%s.%u:%d", myproc.nspace, myproc.rank, 0)) {
            goto done;
        }
        if (0 > asprintf(&keys[1], "BIGTEST:%s.%u:%d", myproc.nspace, myproc.rank, 1)) {
            goto done;
        }
        keys[2] = NULL;
        if (PMIX_SUCCESS != (rc = PMIx_Unpublish(keys, NULL, 0))) {
            fprintf(stderr, "Client ns %s rank %d: PMIx_Unpublish failed: %d\n", myproc.nspace,
                    myproc.rank, rc);
            goto done;
        }
        // purge the rest
        if (PMIX_SUCCESS != (rc = PMIx_Unpublish(NULL, NULL, 0))) {
            fprintf(stderr, "Client ns %s rank %d: PMIx_Unpublish purge failed: %d\n", myproc.nspace,
                    myproc.rank, rc);
            goto done;
        }
        // verify the purge
        if (0 > asprintf(&tmp, "BIGTEST:%s.%u:%d", myproc.nspace, 2, 2)) {
            goto done;
        }
        PMIX_LOAD_KEY(pdata.key, tmp);
        free(tmp);
        // check value
        rc = PMIx_Lookup(&pdata, 1, NULL, 0);
        // should not be found
        if (PMIX_SUCCESS == rc) {
            fprintf(stderr, "Client ns %s rank %d: Purge failed\n",
                    myproc.nspace, myproc.rank);
            goto done;
        } else {
            fprintf(stderr, "Client ns %s rank %d: Purge succeeded: %s\n",
                    myproc.nspace, myproc.rank, PMIx_Error_string(rc));
        }
        // now publish something on different persistences
        firstread = PMIX_PERSIST_PROC;
        PMIX_INFO_LOAD(&info[1], PMIX_PERSISTENCE, &firstread, PMIX_PERSIST);
        if (0 > asprintf(&tmp, "RANGETEST:%s.%u:0", myproc.nspace, myproc.rank)) {
            goto done;
        }
        bcount = 1;
        PMIX_INFO_LOAD(&info[0], tmp, &bcount, PMIX_SIZE);
        free(tmp);
        if (PMIX_SUCCESS != (rc = PMIx_Publish(info, 2))) {
            fprintf(stderr, "Client ns %s rank %d: PMIx_Publish failed: %d\n", myproc.nspace,
                    myproc.rank, rc);
            goto done;
        }
        PMIX_INFO_DESTRUCT(&info[0]);
        firstread = PMIX_PERSIST_APP;
        PMIX_INFO_LOAD(&info[1], PMIX_PERSISTENCE, &firstread, PMIX_PERSIST);
        if (0 > asprintf(&tmp, "RANGETEST:%s.%u:1", myproc.nspace, myproc.rank)) {
            goto done;
        }
        bcount = 2;
        PMIX_INFO_LOAD(&info[0], tmp, &bcount, PMIX_SIZE);
        free(tmp);
        if (PMIX_SUCCESS != (rc = PMIx_Publish(info, 2))) {
            fprintf(stderr, "Client ns %s rank %d: PMIx_Publish failed: %d\n", myproc.nspace,
                    myproc.rank, rc);
            goto done;
        }
        PMIX_INFO_DESTRUCT(&info[0]);
        firstread = PMIX_PERSIST_SESSION;
        PMIX_INFO_LOAD(&info[1], PMIX_PERSISTENCE, &firstread, PMIX_PERSIST);
        if (0 > asprintf(&tmp, "RANGETEST:%s.%u:2", myproc.nspace, myproc.rank)) {
            goto done;
        }
        bcount = 3;
        PMIX_INFO_LOAD(&info[0], tmp, &bcount, PMIX_SIZE);
        free(tmp);
        if (PMIX_SUCCESS != (rc = PMIx_Publish(info, 2))) {
            fprintf(stderr, "Client ns %s rank %d: PMIx_Publish failed: %d\n", myproc.nspace,
                    myproc.rank, rc);
            goto done;
        }
        PMIX_INFO_DESTRUCT(&info[0]);

        // unpublish only those on the app persistence
        if (0 > asprintf(&keys[0], "RANGETEST:%s.%u:1", myproc.nspace, myproc.rank)) {
            goto done;
        }
        keys[1] = NULL;
        firstread = PMIX_PERSIST_APP;
        PMIX_INFO_LOAD(&info[0], PMIX_PERSISTENCE, &firstread, PMIX_PERSIST);
        if (PMIX_SUCCESS != (rc = PMIx_Unpublish(keys, info, 1))) {
            fprintf(stderr, "Client ns %s rank %d: PMIx_Unpublish of app persistence failed: %s\n", myproc.nspace,
                    myproc.rank, PMIx_Error_string(rc));
            goto done;
        } else {
            fprintf(stderr, "Client ns %s rank %d: PMIx_Unpublish of app persistence succeeded\n", myproc.nspace,
                    myproc.rank);
        }

        // verify the app persistence keys are gone
        PMIX_PDATA_CONSTRUCT(&pdata);
        if (0 > asprintf(&tmp, "RANGETEST:%s.%u:1", myproc.nspace, myproc.rank)) {
            goto done;
        }
        PMIX_LOAD_KEY(pdata.key, tmp);
        free(tmp);
        // check value
        rc = PMIx_Lookup(&pdata, 1, NULL, 0);
        if (PMIX_SUCCESS == rc) {
            fprintf(stderr, "Client ns %s rank %d: PMIx_Lookup succeeded instead of failed\n", myproc.nspace,
                    myproc.rank);
            goto done;
        } else {
            fprintf(stderr, "Client ns %s rank %d: PMIx_Lookup of app persistence value correctly failed: %s\n", myproc.nspace,
                    myproc.rank, PMIx_Error_string(rc));
        }

        PMIX_PDATA_DESTRUCT(&pdata);

        // verify the session persistence key is still there
        PMIX_PDATA_CONSTRUCT(&pdata);
        if (0 > asprintf(&tmp, "RANGETEST:%s.%u:2", myproc.nspace, myproc.rank)) {
            goto done;
        }
        PMIX_LOAD_KEY(pdata.key, tmp);
        free(tmp);
        // check value
        rc = PMIx_Lookup(&pdata, 1, NULL, 0);
        if (PMIX_SUCCESS != rc) {
            fprintf(stderr, "Client ns %s rank %d: PMIx_Lookup failed: %s\n", myproc.nspace,
                    myproc.rank, PMIx_Error_string(rc));
            goto done;
        } else {
            fprintf(stderr, "Client ns %s rank %d: PMIx_Lookup of session persistence value succeeded\n", myproc.nspace,
                    myproc.rank);
        }
        PMIX_PDATA_DESTRUCT(&pdata);

    }

done:
    /* finalize us */
    if (PMIX_SUCCESS != (rc = PMIx_Finalize(NULL, 0))) {
        fprintf(stderr, "Client ns %s rank %d:PMIx_Finalize failed: %d\n", myproc.nspace,
                myproc.rank, rc);
    } else {
        fprintf(stderr, "Client ns %s rank %d:PMIx_Finalize successfully completed\n",
                myproc.nspace, myproc.rank);
    }
    fflush(stderr);
    return (0);
}

File: examples/python/pubstress.py
#!/usr/bin/env python3
#
# Copyright (c) 2026      Nanook Consulting  All rights reserved.
# $COPYRIGHT$
#
# Additional copyrights may follow
#
# $HEADER$
#
# Python port of examples/pubstress.c
#
# A heavier workout of the publish/lookup service:
#
#   ranks 0,1   ping-pong PMIX_PERSIST_FIRST_READ data, as pub2 does, but
#               with deliberate delays so the two get out of step
#   rank 2      publishes a value each iteration that everyone else reads
#               (no first-read persistence, so many readers can see it)
#   rank 2      then exercises unpublish: by key, a full purge, and a
#               persistence-qualified unpublish that must remove only the
#               app-scoped key and leave the session-scoped one
#
# Takes an optional iteration count; requires at least 4 processes.

import os
import sys
import time

from examples import *

NITER = 10


def main():
    if 1 < len(sys.argv):
        iters = int(sys.argv[1])
    else:
        iters = NITER

    client = PMIxClient()

    # init us
    rc, myproc = client.init(None)
    if PMIX_SUCCESS != rc:
        eprint("Client: PMIx_Init failed: %d" % rc)
        sys.exit(0)

    # get our job size
    proc = {'nspace': myproc['nspace'], 'rank': PMIX_RANK_WILDCARD}
    rc, val = client.get(proc, PMIX_JOB_SIZE, None)
    if PMIX_SUCCESS != rc:
        eprint("Client ns %s rank %d: PMIx_Get job size failed: %d"
               % (myproc['nspace'], myproc['rank'], rc))
        return done(client, myproc)
    nprocs = val['value']
    # require at least 4 procs
    if nprocs < 4:
        if 0 == myproc['rank']:
            eprint("%s requires at least 4 processes"
                   % os.path.basename(sys.argv[0]))
        sys.exit(1)

    persist = {'key': PMIX_PERSISTENCE, 'value': PMIX_PERSIST_FIRST_READ,
               'val_type': PMIX_PERSIST}
    lkinfo = [{'key': PMIX_WAIT, 'value': True, 'val_type': PMIX_BOOL},
              {'key': PMIX_TIMEOUT, 'value': 10, 'val_type': PMIX_INT}]

    if myproc['rank'] < 2:
        for n in range(iters):
            if 0 == myproc['rank']:
                if n % 2:   # delay us a bit so other ranks get ahead
                    time.sleep(1)
                # publish something
                key = "FOOBAR:%s.%u:%d" % (myproc['nspace'],
                                           myproc['rank'], n)
                rc = client.publish([{'key': key, 'value': n,
                                      'val_type': PMIX_INT}, persist])
                if PMIX_SUCCESS != rc:
                    eprint("Client ns %s rank %d: PMIx_Publish failed: %d"
                           % (myproc['nspace'], myproc['rank'], rc))
                    return done(client, myproc)

                # lookup other rank's value
                key = "BAZ:%s.%u:%d" % (myproc['nspace'], 1, n)
                rc, pdata = client.lookup([{'key': key}], lkinfo)
                if PMIX_SUCCESS != rc:
                    eprint("Client ns %s rank %d: PMIx_Lookup failed: %d"
                           % (myproc['nspace'], myproc['rank'], rc))
                    return done(client, myproc)
                if not check(client, myproc, pdata[0], 1, PMIX_INT, n):
                    return done(client, myproc)
                eprint("PUBLISH-LOOKUP SUCCEEDED: %d" % n)
            elif 1 == myproc['rank']:
                if 0 == n:
                    time.sleep(1)
                # lookup other rank's value
                key = "FOOBAR:%s.%u:%d" % (myproc['nspace'], 0, n)
                rc, pdata = client.lookup([{'key': key}], lkinfo)
                if PMIX_SUCCESS != rc:
                    eprint("Client ns %s rank %d: PMIx_Lookup failed: %d"
                           % (myproc['nspace'], myproc['rank'], rc))
                    return done(client, myproc)
                if not check(client, myproc, pdata[0], 0, PMIX_INT, n):
                    return done(client, myproc)
                # publish something
                key = "BAZ:%s.%u:%d" % (myproc['nspace'],
                                        myproc['rank'], n)
                rc = client.publish([{'key': key, 'value': n,
                                      'val_type': PMIX_INT}, persist])
                if PMIX_SUCCESS != rc:
                    eprint("Client ns %s rank %d: PMIx_Publish failed: %d"
                           % (myproc['nspace'], myproc['rank'], rc))
                    return done(client, myproc)
        return done(client, myproc)

    # we cannot participate in the "firstread" cases as only
    # one other proc can read the value
    for n in range(iters):
        if 2 == myproc['rank']:
            if n % 2:   # delay us a bit so other ranks get ahead
                time.sleep(1)
            # publish something
            key = "BIGTEST:%s.%u:%d" % (myproc['nspace'], myproc['rank'], n)
            bcount = 1234 + n
            eprint("Client ns %s rank %d: Published: %lu"
                   % (myproc['nspace'], myproc['rank'], bcount))
            rc = client.publish([{'key': key, 'value': bcount,
                                  'val_type': PMIX_SIZE}])
            if PMIX_SUCCESS != rc:
                eprint("Client ns %s rank %d: PMIx_Publish failed: %d"
                       % (myproc['nspace'], myproc['rank'], rc))
                return done(client, myproc)
        else:
            # everyone else retrieves it
            key = "BIGTEST:%s.%u:%d" % (myproc['nspace'], 2, n)
            rc, pdata = client.lookup([{'key': key}], lkinfo)
            if PMIX_SUCCESS != rc:
                eprint("Client ns %s rank %d: PMIx_Lookup failed: %d"
                       % (myproc['nspace'], myproc['rank'], rc))
                return done(client, myproc)
            if not check(client, myproc, pdata[0], 2, PMIX_SIZE, 1234 + n):
                return done(client, myproc)
            eprint("Client ns %s rank %d: PUBLISH-LOOKUP BIGTEST "
                   "SUCCEEDED: %d" % (myproc['nspace'], myproc['rank'], n))

    rc = client.fence([proc], None)

    if 2 == myproc['rank']:
        keys = ["BIGTEST:%s.%u:%d" % (myproc['nspace'], myproc['rank'], 0),
                "BIGTEST:%s.%u:%d" % (myproc['nspace'], myproc['rank'], 1)]
        rc = client.unpublish(keys, None)
        if PMIX_SUCCESS != rc:
            eprint("Client ns %s rank %d: PMIx_Unpublish failed: %d"
                   % (myproc['nspace'], myproc['rank'], rc))
            return done(client, myproc)
        # purge the rest
        rc = client.unpublish(None, None)
        if PMIX_SUCCESS != rc:
            eprint("Client ns %s rank %d: PMIx_Unpublish purge failed: %d"
                   % (myproc['nspace'], myproc['rank'], rc))
            return done(client, myproc)
        # verify the purge - should not be found
        key = "BIGTEST:%s.%u:%d" % (myproc['nspace'], 2, 2)
        rc, pdata = client.lookup([{'key': key}], None)
        if PMIX_SUCCESS == rc:
            eprint("Client ns %s rank %d: Purge failed"
                   % (myproc['nspace'], myproc['rank']))
            return done(client, myproc)
        eprint("Client ns %s rank %d: Purge succeeded: %s"
               % (myproc['nspace'], myproc['rank'], client.error_string(rc)))

        # now publish something on different persistences
        for idx, (persistence, value) in enumerate(
                ((PMIX_PERSIST_PROC, 1), (PMIX_PERSIST_APP, 2),
                 (PMIX_PERSIST_SESSION, 3))):
            key = "RANGETEST:%s.%u:%d" % (myproc['nspace'],
                                          myproc['rank'], idx)
            rc = client.publish([{'key': key, 'value': value,
                                  'val_type': PMIX_SIZE},
                                 {'key': PMIX_PERSISTENCE,
                                  'value': persistence,
                                  'val_type': PMIX_PERSIST}])
            if PMIX_SUCCESS != rc:
                eprint("Client ns %s rank %d: PMIx_Publish failed: %d"
                       % (myproc['nspace'], myproc['rank'], rc))
                return done(client, myproc)

        # unpublish only those on the app persistence
        keys = ["RANGETEST:%s.%u:1" % (myproc['nspace'], myproc['rank'])]
        info = [{'key': PMIX_PERSISTENCE, 'value': PMIX_PERSIST_APP,
                 'val_type': PMIX_PERSIST}]
        rc = client.unpublish(keys, info)
        if PMIX_SUCCESS != rc:
            eprint("Client ns %s rank %d: PMIx_Unpublish of app persistence "
                   "failed: %s" % (myproc['nspace'], myproc['rank'],
                                   client.error_string(rc)))
            return done(client, myproc)
        eprint("Client ns %s rank %d: PMIx_Unpublish of app persistence "
               "succeeded" % (myproc['nspace'], myproc['rank']))

        # verify the app persistence keys are gone
        key = "RANGETEST:%s.%u:1" % (myproc['nspace'], myproc['rank'])
        rc, pdata = client.lookup([{'key': key}], None)
        if PMIX_SUCCESS == rc:
            eprint("Client ns %s rank %d: PMIx_Lookup succeeded instead of "
                   "failed" % (myproc['nspace'], myproc['rank']))
            return done(client, myproc)
        eprint("Client ns %s rank %d: PMIx_Lookup of app persistence value "
               "correctly failed: %s" % (myproc['nspace'], myproc['rank'],
                                         client.error_string(rc)))

        # verify the session persistence key is still there
        key = "RANGETEST:%s.%u:2" % (myproc['nspace'], myproc['rank'])
        rc, pdata = client.lookup([{'key': key}], None)
        if PMIX_SUCCESS != rc:
            eprint("Client ns %s rank %d: PMIx_Lookup failed: %s"
                   % (myproc['nspace'], myproc['rank'],
                      client.error_string(rc)))
            return done(client, myproc)
        eprint("Client ns %s rank %d: PMIx_Lookup of session persistence "
               "value succeeded" % (myproc['nspace'], myproc['rank']))

    return done(client, myproc)


def check(client, myproc, pdata, rank, val_type, value):
    """Check a lookup return for value and source."""
    if myproc['nspace'] != pdata['proc']['nspace']:
        eprint("Client ns %s rank %d: PMIx_Lookup returned wrong nspace: %s"
               % (myproc['nspace'], myproc['rank'], pdata['proc']['nspace']))
        return False
    if rank != pdata['proc']['rank']:
        eprint("Client ns %s rank %d: PMIx_Lookup returned wrong rank: %d"
               % (myproc['nspace'], myproc['rank'], pdata['proc']['rank']))
        return False
    if val_type != pdata['val_type']:
        eprint("Client ns %s rank %d: PMIx_Lookup returned wrong type: %s"
               % (myproc['nspace'], myproc['rank'],
                  client.data_type_string(pdata['val_type'])))
        return False
    if value != pdata['value']:
        eprint("Client ns %s rank %d: PMIx_Lookup returned wrong value: %lu"
               % (myproc['nspace'], myproc['rank'], pdata['value']))
        return False
    return True


def done(client, myproc):
    # finalize us
    rc = client.finalize(None)
    if PMIX_SUCCESS != rc:
        eprint("Client ns %s rank %d:PMIx_Finalize failed: %d"
               % (myproc['nspace'], myproc['rank'], rc))
    else:
        eprint("Client ns %s rank %d:PMIx_Finalize successfully completed"
               % (myproc['nspace'], myproc['rank']))
    return 0


if __name__ == '__main__':
    sys.exit(main())
